### PR TITLE
Change to owned LoggerProvider

### PR DESCRIPTION
**Pull request by hdost**
The interface to an item shouldn't take an inner value since it's considered inner, this also allows for further optimizations in the future as it hides the complexity from the user.

Relates #1209

## Changes

Please provide a brief description of the changes here.

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by lalitb** [comment]
Trying to understand more here - so with this PR, each logger holds its own complete instance of LoggerProvider, which would be different from the other loggers, and also different from the global LoggerProvider. And will calling shutdown on the global loggerProvider ensure that the logs through these loggers are not exported ?

Also, would there be increased memory usage if many Logger instances are created, as each carries a full copy of the LoggerProvider and its internal state.

**Comment by hdost** [comment]
> Trying to understand more here - so with this PR, each logger holds its own complete instance of LoggerProvider, which would be different from the other loggers, and also different from the global LoggerProvider. And will calling shutdown on the global loggerProvider ensure that the logs through these loggers are not exported ?

The inner is still the same, so it should, I'll double check that there's a test. 

> Also, would there be increased memory usage if many Logger instances are created, as each carries a full copy of the LoggerProvider and its internal state.

It won't be a full copy, the inner is an Arc, so the outer would be an additional pointer memory wise. 

> 
> 
> Also, please change the PR title to reflect LoggerProvider instead of TracerProvider :)

Done 😅

File: opentelemetry-sdk/src/logs/log_emitter.rs
Comment (TommyCpp) [inline opentelemetry-sdk/src/logs/log_emitter.rs:55]: nit: `self.clone()` probably cleaner IMO

**Review by TommyCpp — APPROVED** [review]

**Comment by lalitb** [comment]
> It won't be a full copy, the inner is an Arc, so the outer would be an additional pointer memory wise.

Ah good point. I didn't realize that inner is Arc type, and since there is no `owned` trait implementation, the `to_owned()` will actually do `clone`, an so all outer loggerprovider instances would share the single `inner` instance.

**Comment by lalitb** [comment]
> And will calling shutdown on the global loggerProvider ensure that the logs through these loggers are not exported ?

It would be good to validate this use-case. It seems [shutdown_logger_provider](https://github.com/open-telemetry/opentelemetry-rust/blob/f9b08c10809e3b39497e21cccaa0d0c49b0216dc/opentelemetry/src/global/logs.rs#L134) will invoke shutdown on the LogProcessors when they are are [dropped](https://github.com/open-telemetry/opentelemetry-rust/blob/f9b08c10809e3b39497e21cccaa0d0c49b0216dc/opentelemetry-sdk/src/logs/log_emitter.rs#L116). And now if any of the loggers have it's reference in separate thread, the shutdown will not be invoked.  One option can be to add a new `LoggerProvider::Shutdown()` method, and explicitly invoke it from within `shutdown_logger_provider` . This new method will invoke shutdown on all processors, and so ensure that all the existing logs are flushed. And no new logs can be exported with processors in shutdown state.

**Comment by TommyCpp** [comment]
Hmm I think this PR is the same as #1446 ?

**Comment by hdost** [comment]
> Hmm I think this PR is the same as #1446 ?

Yes it's meant to be an alternative because right now the initial interface requires passing the under which is inherently leaky as an interface. This aims to remove the necessity.

**Comment by lalitb** [comment]
Surprisingly, this PR has better improvement over #1446, as visible from the stress test:
```bash
/home/labhas/opentelemetry-rust/stress$ cargo run --bin logs --release
```

#1455 (current PR):
Number of threads: 8
Throughput: 20,641,400 iterations/sec
Throughput: 20,336,600 iterations/sec

#1446:
Number of threads: 8
Throughput: 14,935,400 iterations/sec
Throughput: 15,633,600 iterations/sec

#main branch:
Number of threads: 8
Throughput: 11,473,200 iterations/sec
Throughput: 11,909,200 iterations/sec

**Comment by hdost** [comment]
That's a nice benefit as well. I believe it's because we're really doing less things in this one. 


Side note: I really would love to run whatever machines y'all are running because my same stress benchmarks never get as good of results. Maybe after the next refresh for my laptop 😅

**Comment by lalitb** [comment]
> > And will calling shutdown on the global loggerProvider ensure that the logs through these loggers are not exported ?
> 
> It would be good to validate this use-case. It seems [shutdown_logger_provider](https://github.com/open-telemetry/opentelemetry-rust/blob/f9b08c10809e3b39497e21cccaa0d0c49b0216dc/opentelemetry/src/global/logs.rs#L134) will invoke shutdown on the LogProcessors when the inner instance is [dropped](https://github.com/open-telemetry/opentelemetry-rust/blob/f9b08c10809e3b39497e21cccaa0d0c49b0216dc/opentelemetry-sdk/src/logs/log_emitter.rs#L116). And now if any of the loggers have it's reference in separate thread, the shutdown will not be invoked. One option can be to add a new `LoggerProvider::Shutdown()` method, and explicitly invoke it from within `shutdown_logger_provider` . This new method will invoke shutdown on all processors, and so ensure that all the existing logs are flushed. And no new logs can be exported with processors in shutdown state.

Thinking about this scenario again,  it seems that the potential issue of leaking providers might not be as significant if we can document it clearly to explain the risk. Specifically, providers could be leaked if any of the tasks indefinitely retain their reference, which can result in a shutdown not triggering.

**Review by lalitb — APPROVED** [review]

**Comment by cijothomas** [comment]
> > Hmm I think this PR is the same as #1446 ?
> 
> Yes it's meant to be an alternative because right now the initial interface requires passing the under which is inherently leaky as an interface. This aims to remove the necessity.

Just catching up on PRs! Could you describe any pros/cons of this vs the alternate in the PR desc?  Also, if feasible, can you add the stress test numbers you are observing. I am not seeing the same gains as Lalit has posted above, so wondering what is the actual gain! (i'll try to run more tests and post back as well soon)

**Comment by lalitb** [comment]
My earlier results were on the Azure dev box, but I can see perf improvement on my Windows desktop with WSL2:

#main branch:
     Running `/home/labhas/obs/ot/opentelemetry-rust/target/release/logs`
Number of threads: 10
Throughput: 9,659,000 iterations/sec
Throughput: 9,166,800 iterations/sec
Throughput: 9,068,400 iterations/sec
Throughput: 9,013,600 iterations/sec

#this PR:
     Running `/home/labhas/obs/ot/opentelemetry-rust/target/release/logs`
Number of threads: 10
Throughput: 28,168,200 iterations/sec
Throughput: 26,013,800 iterations/sec
Throughput: 23,196,200 iterations/sec
Throughput: 24,413,400 iterations/sec

**Comment by hdost** [comment]
Are we still thinking that we need to have the testing before this gets merged?

I'm trying to thing what would be a valid testing scenario that we'd want to run.

**Comment by cijothomas** [comment]
> Are we still thinking that we need to have the testing before this gets merged?
> 
> I'm trying to thing what would be a valid testing scenario that we'd want to run.

From SIG Call:
We don't need to block the PR for tests - a targetted test can be added as a followup.
Scenario: https://github.com/open-telemetry/opentelemetry-rust/pull/1455#issuecomment-1877690239 
There are other issues reporting about shutdown scenarios, maybe we need a comprehensive testing targetting shutdowns, separate from this PR.

File: opentelemetry/CHANGELOG.md
Comment (cijothomas) [inline opentelemetry/CHANGELOG.md:31]:
```suggestion
  [1455](https://github.com/open-telemetry/opentelemetry-rust/pull/1455)
```

**Review by cijothomas — APPROVED** [review]
LGTM. 
Maybe good to add changelog entry to the sdk.
We can definitely follow up to do targeted tests for shutdown part.

**Comment by cijothomas** [comment]
I ran stress test for logs after modifying the no-op processor to [return false ](https://github.com/open-telemetry/opentelemetry-rust/blob/main/stress/src/logs.rs#L61) for event_enabled.
main branch:

Number of threads: 8
Throughput: 75,537,000 iterations/sec
Throughput: 77,545,600 iterations/sec

This pr:

Number of threads: 8
Throughput: 579,008,400 iterations/sec
Throughput: 586,900,800 iterations/sec

On the other hand, if using a [no-op loglayer](https://github.com/open-telemetry/opentelemetry-rust/blob/main/stress/src/logs.rs#L18) with event_enabled returning false, the same test gives below:

Number of threads: 8
Throughput: 1,106,156,200 iterations/sec
Throughput: 1,069,867,600 iterations/sec

This test shows that OTel still has a lot more room for further optimizations. Those can be explored in the future. This PR already boosts the perf significantly!

**Pull request merged**
